spec/unit/matrixrtc/MembershipManager.spec.ts

```diff
@@ -0,0 +1,611 @@
+ */
+/*
+Copyright 2025 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { type Mock } from "jest-mock";
+import waitForExpect from "wait-for-expect";
+
+import { EventType, HTTPError, MatrixError, type Room } from "../../../src";
+import { type Focus, type LivekitFocusActive, type SessionMembershipData } from "../../../src/matrixrtc";
+import { LegacyMembershipManager } from "../../../src/matrixrtc/MembershipManager";
+import { makeMockClient, makeMockRoom, membershipTemplate, mockCallMembership, type MockClient } from "./mocks";
+import { flushPromises } from "../../test-utils/flushPromises";
+// import { MembershipManager } from "../../../src/matrixrtc/NewMembershipManager";
+
+function waitForMockCall(method: any, returnVal?: any) {
+    return new Promise<void>((resolve) => {
+        (method as Mock).mockImplementation(() => {
+            resolve();
+            return returnVal;
+        });
+    });
+}
+
+function createAsyncHandle(method: any) {
+    const handle: { resolve?: (...args: unknown[]) => void; reject?: (...args: any[]) => void } = {};
+    (method as Mock).mockImplementation(() => {
+        return new Promise((resolve, reject) => {
+            handle.reject = reject;
+            handle.resolve = resolve;
+        });
+    });
+    return handle;
+}
+
+/**
+ * Tests different MembershipManager implementations. Some tests don't apply to `LegacyMembershipManager`
+ * use !FailsForLegacy to skip those. See: testEnvironment for more details.
+ */
+describe("MembershipManager", () => {
+    describe.each([
+        { TestMembershipManager: LegacyMembershipManager, description: "LegacyMembershipManager" },
+        // Here we will add the new implementation of the MembershipManager.
+        // It is not yet tested since it would currently fail all tests. Adding the MembershipManger looks like this:
+        // { TestMembershipManager: MembershipManager, description: "MembershipManager" },
+    ])("$description", ({ TestMembershipManager }) => {
+        let client: MockClient;
+        let room: Room;
+        const focusActive: LivekitFocusActive = {
+            focus_selection: "oldest_membership",
+            type: "livekit",
+        };
+        const focus: Focus = {
+            type: "livekit",
+            livekit_service_url: "https://active.url",
+            livekit_alias: "!active:active.url",
+        };
+
+        beforeEach(() => {
+            // default to fake timers
```

Comment:
Warning: there's a good chance someone won't notice this is happening and it will bite them in weird ways because something that uses a short setTimeout delay won't run automatically.

Reply:
The whole file will only use FakeTimers. So I think there are enough examples with `jest.advacenTimers...` so that will quickly resolve?